File: foxglove-extension/src/dewarp.ts
/**
 * Cylindrical camera model used by these vehicles' cameras:
 *
 *   theta = atan2(X, Z)                 // horizontal angle from optical axis
 *   r     = sqrt(X*X + Z*Z)
 *   u     = fx * theta + cx
 *   v     = fy * (Y / r) + cy
 *
 * for a 3D point (X, Y, Z) in the camera frame (Z forward). This is not a
 * pinhole/plumb_bob model, so foxglove.CameraCalibration's distortion
 * pipeline doesn't apply. To display these images undistorted, we build a
 * remap lookup table: for every destination (rectified, pinhole) pixel we
 * compute the camera-frame ray a pinhole camera with virtual focal length
 * `fVirtual` would have produced, then invert the cylindrical projection
 * above to find the corresponding source pixel.
 */

export interface CylindricalIntrinsics {
  fx: number;
  fy: number;
  cx: number;
  cy: number;
  width: number;
  height: number;
}

export interface RemapLut {
  width: number;
  height: number;
  // For each destination pixel i, srcX[i]/srcY[i] hold the source pixel
  // coordinates to sample (bilinearly). NaN entries are out of bounds.
  srcX: Float32Array;
  srcY: Float32Array;
}

/**
 * Build a remap LUT that dewarps `intrinsics`-shaped cylindrical source
 * images into a `destWidth`x`destHeight` pinhole-equivalent perspective view
 * with the given horizontal field of view (degrees).
 */
export function buildRemapLut(
  intrinsics: CylindricalIntrinsics,
  destWidth: number,
  destHeight: number,
  horizontalFovDeg: number,
): RemapLut {
  const srcX = new Float32Array(destWidth * destHeight);
  const srcY = new Float32Array(destWidth * destHeight);

  const halfFovRad = (horizontalFovDeg * Math.PI) / 180 / 2;
  // Virtual pinhole focal length chosen so the requested horizontal FOV maps
  // across destWidth pixels: tan(halfFov) = (destWidth/2) / fVirtual.
  const fVirtual = destWidth / 2 / Math.tan(halfFovRad);
  const destCx = destWidth / 2;
  const destCy = destHeight / 2;

  for (let dy = 0; dy < destHeight; dy++) {
    for (let dx = 0; dx < destWidth; dx++) {
      const X = (dx - destCx) / fVirtual;
      const Y = (dy - destCy) / fVirtual;
      const Z = 1;

      const theta = Math.atan2(X, Z);
      const r = Math.sqrt(X * X + Z * Z);

      const u = intrinsics.fx * theta + intrinsics.cx;
      const v = (intrinsics.fy * Y) / r + intrinsics.cy;

      const idx = dy * destWidth + dx;
      if (u >= 0 && u < intrinsics.width && v >= 0 && v < intrinsics.height) {
        srcX[idx] = u;
        srcY[idx] = v;
      } else {
        srcX[idx] = NaN;
        srcY[idx] = NaN;
      }
    }
  }

  return { width: destWidth, height: destHeight, srcX, srcY };
}

/** Bilinearly sample `src` (RGBA, width x height) at floating point (x, y). */
function sampleBilinear(
  src: Uint8ClampedArray,
  srcWidth: number,
  srcHeight: number,
  x: number,
  y: number,
  out: [number, number, number, number],
): void {
  const x0 = Math.floor(x);
  const y0 = Math.floor(y);
  const x1 = Math.min(x0 + 1, srcWidth - 1);
  const y1 = Math.min(y0 + 1, srcHeight - 1);
  const fx = x - x0;
  const fy = y - y0;

  for (let c = 0; c < 4; c++) {
    const p00 = src[(y0 * srcWidth + x0) * 4 + c] ?? 0;
    const p10 = src[(y0 * srcWidth + x1) * 4 + c] ?? 0;
    const p01 = src[(y1 * srcWidth + x0) * 4 + c] ?? 0;
    const p11 = src[(y1 * srcWidth + x1) * 4 + c] ?? 0;
    const top = p00 * (1 - fx) + p10 * fx;
    const bottom = p01 * (1 - fx) + p11 * fx;
    out[c] = top * (1 - fy) + bottom * fy;
  }
}

/** Apply a remap LUT to a decoded source image, producing a dewarped ImageData. */
export function applyRemap(
  lut: RemapLut,
  srcImageData: ImageData,
): ImageData {
  const dest = new ImageData(lut.width, lut.height);
  const pixel: [number, number, number, number] = [0, 0, 0, 0];

  for (let i = 0; i < lut.width * lut.height; i++) {
    const sx = lut.srcX[i];
    const sy = lut.srcY[i];
    if (sx === undefined || sy === undefined || Number.isNaN(sx) || Number.isNaN(sy)) {
      dest.data[i * 4 + 3] = 0; // transparent / out of frame
      continue;
    }
    sampleBilinear(srcImageData.data, srcImageData.width, srcImageData.height, sx, sy, pixel);
    dest.data[i * 4 + 0] = pixel[0];
    dest.data[i * 4 + 1] = pixel[1];
    dest.data[i * 4 + 2] = pixel[2];
    dest.data[i * 4 + 3] = 255;
  }

  return dest;
}

File: folder_to_mcap/dewarp.py
"""Python port of the cylindrical-camera dewarp math used by the companion
Foxglove extension (foxglove-extension/src/dewarp.ts).

The converter uses this to also emit a rectified (virtual pinhole) image
alongside each camera's raw cylindrical frames, plus a matching
foxglove.CameraCalibration message, so Foxglove's built-in 3D panel can
project the images directly (it only understands pinhole/plumb_bob
calibration, not this vehicle's cylindrical projection).

    theta = atan2(X, Z)
    r     = sqrt(X*X + Z*Z)
    u     = fx * theta + cx
    v     = fy * (Y / r) + cy

for a 3D point (X, Y, Z) in the camera frame (Z forward). To dewarp, for
every destination pixel we compute the camera-frame ray a virtual pinhole
camera would have produced, then invert the projection above to find the
corresponding source pixel to sample.
"""

from __future__ import annotations

import math

import numpy as np


class RemapLut:
    def __init__(self, u: np.ndarray, v: np.ndarray, valid: np.ndarray, f_virtual: float, dest_width: int, dest_height: int):
        self.u = u
        self.v = v
        self.valid = valid
        self.f_virtual = f_virtual
        self.dest_width = dest_width
        self.dest_height = dest_height


def build_remap_lut(
    fx: float,
    fy: float,
    cx: float,
    cy: float,
    src_width: int,
    src_height: int,
    dest_width: int,
    dest_height: int,
    horizontal_fov_deg: float,
) -> RemapLut:
    """Build a remap LUT that dewarps a cylindrical source image (described by
    fx/fy/cx/cy/src_width/src_height) into a dest_width x dest_height
    pinhole-equivalent perspective view with the given horizontal FOV."""
    half_fov = math.radians(horizontal_fov_deg) / 2
    # Virtual pinhole focal length chosen so the requested horizontal FOV maps
    # across dest_width pixels: tan(halfFov) = (destWidth/2) / fVirtual.
    f_virtual = dest_width / 2 / math.tan(half_fov)
    dest_cx = dest_width / 2
    dest_cy = dest_height / 2

    dx, dy = np.meshgrid(
        np.arange(dest_width, dtype=np.float64), np.arange(dest_height, dtype=np.float64)
    )
    X = (dx - dest_cx) / f_virtual
    Y = (dy - dest_cy) / f_virtual
    Z = 1.0

    theta = np.arctan2(X, Z)
    r = np.sqrt(X * X + Z * Z)

    u = fx * theta + cx
    v = fy * Y / r + cy

    valid = (u >= 0) & (u < src_width) & (v >= 0) & (v < src_height)
    return RemapLut(u, v, valid, f_virtual, dest_width, dest_height)


def apply_remap(lut: RemapLut, src_rgb: np.ndarray) -> np.ndarray:
    """Bilinearly resample src_rgb (H, W, 3 uint8) per `lut`, returning a
    (dest_height, dest_width, 3 uint8) array. Pixels the LUT marked
    out-of-bounds are filled black."""
    src_h, src_w = src_rgb.shape[:2]
    u, v = lut.u, lut.v

    u0 = np.floor(u)
    v0 = np.floor(v)
    fu = (u - u0).astype(np.float32)[..., None]
    fv = (v - v0).astype(np.float32)[..., None]

    u0c = np.clip(u0, 0, src_w - 1).astype(np.int64)
    v0c = np.clip(v0, 0, src_h - 1).astype(np.int64)
    u1c = np.clip(u0 + 1, 0, src_w - 1).astype(np.int64)
    v1c = np.clip(v0 + 1, 0, src_h - 1).astype(np.int64)

    p00 = src_rgb[v0c, u0c].astype(np.float32)
    p10 = src_rgb[v0c, u1c].astype(np.float32)
    p01 = src_rgb[v1c, u0c].astype(np.float32)
    p11 = src_rgb[v1c, u1c].astype(np.float32)

    top = p00 * (1 - fu) + p10 * fu
    bottom = p01 * (1 - fu) + p11 * fu
    out = top * (1 - fv) + bottom * fv
    out = np.clip(out, 0, 255).astype(np.uint8)
    out[~lut.valid] = 0
    return out
